Bmm sparse dense

kurtamohler opened the pull request:
Add sparse-dense BMM operation for CUDA and CPU.

Closes #5672

kurtamohler committed:
Add CPU implementation of bmm() for sparse times dense matrices

Issue #5672

kurtamohler committed:
Improve CPU sparse BMM performance by mimicking sparse MM

Issue #5672

kurtamohler committed:
Use un-batched sparse x dense operation underneath BMM

Issue #5672

kurtamohler committed:
Improve sparse x dense BMM with binary search for matrix elements

Issue #5672

kurtamohler committed:
Remove unnecessary headers

Issue #5672

kurtamohler committed:
Move contents of sparse MM dispatch to separate function

This will allow reuse of the same code in BMM

kurtamohler committed:
Add CUDA version of sparse x dense BMM

Does not perform well, needs optimization
Issue #5672

kurtamohler committed:
Use CUDA binary search to find each matrix start point

kurtamohler committed:
Update CUDA sparse BMM to use cusparseSpMM

Does not give the correct result yet

Issue #5672

kurtamohler committed:
Use 32-bit indices in cusparseSpMM to get correct results

Issue #5672

kurtamohler committed:
Avoid unnecessarily zeroing the result tensor

kurtamohler committed:
Cleanup and consolidate code

kurtamohler committed:
Destroy cusparse matrix descriptors after multiplication

kurtamohler committed:
Remove unused tensor

kurtamohler committed:
Improve performance by avoiding tensor indexing and slicing

Issue #5672

kurtamohler committed:
Handle nnz=0 cases in CPU BMM

kurtamohler committed:
Handle nnz=0 cases in GPU BMM

kurtamohler committed:
Add unit test for sparse-dense BMM

Issue #5672

kurtamohler committed:
Add deterministic option to cuda sparse BMM

Add new test for deterministic arg, and fix base test to
properly run cuda or cpu version

Issue #5672

kurtamohler committed:
Update BMM documentation to explain deterministic flag

Issue #5672

kurtamohler committed:
Fix deterministic arg to use correct alg

kurtamohler committed:
Use pointer type indexing rather than void pointer arithmetic

kurtamohler commented:
The clang-tidy check is failing to install clang-tidy:
```
+ sudo apt-get install -y clang-tidy-8
Reading package lists...
Building dependency tree...
Reading state information...
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 clang-tidy-8 : Depends: libllvm8 (= 1:8.0.1+svn369350-1~exp1~20200112113617.82) but 1:8.0.1+svn369350-1~exp1~20200114191400.80 is to be installed
                Depends: clang-tools-8 but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
##[error]Process completed with exit code 100.
```

kurtamohler commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -708,4 +714,271 @@ Tensor _sparse_sum_backward_cuda(const Tensor& grad_, const SparseTensor& input_
+
+  // TODO: I'm not sure if this is necessary or not. Does at::resize_()
+  //    ensure contiguity?
+  if (!result.is_contiguous()) {
```
Does anyone know if `at::resize_()` ensures contiguity or not? If it does, I can remove this check.

ezyang replied:
No, `resize_` does not ensure contiguity. Specifically, if you have a non-contiguous tensor of exactly the right size, `resize_` will no op. However, please also note that if you are implementing the `out=` variant of a function, you cannot just call contiguous on it, that will not work. You will need to do the computation into a contiguous out buffer, then do a copy into the non-contiguous preallocated memory.

kurtamohler replied:
Ah ok, I'll do that.

kurtamohler replied:
Alright, what I chose to do is use a separate temporary buffer only if the given one is not contiguous. If it is already contiguous, then we can avoid the performance hit of having to copy the result.

ezyang replied:
Yep, that sounds right to me.

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -637,15 +637,41 @@
+    SparseCUDA: bmm_sparse_cuda
+  supports_named_tensor: True
+
+- func: bmm.deterministic(Tensor self, Tensor mat2, bool deterministic) -> Tensor
```
How about let's make the deterministic argument kwarg only. So `Tensor mat2, *, bool deterministic=False`?

kurtamohler replied:
That means I'd have to have a TensorOptions argument in my C++ implementation right? Do you think I should remove the `bmm.deterministic` version and just add TensorOptions to every bmm function?

kurtamohler replied:
Or maybe I could make it an c10::optional argument, like `Tensor mat2, *, bool? deterministic=False`.

kurtamohler replied:
When I change this line to `- func: bmm.deterministic(Tensor self, Tensor mat2, *, bool deterministic=False) -> Tensor` and rebuild, I get this error:

```
Writing torch/csrc/autograd/generated/python_variable_methods.cpp
Traceback (most recent call last):
  File "tools/setup_helpers/generate_code.py", line 93, in <module>
    main()
  File "tools/setup_helpers/generate_code.py", line 88, in main
    options.selected_op_list_path,
  File "tools/setup_helpers/generate_code.py", line 43, in generate_code
    gen_autograd_python(declarations_path or DECLARATIONS_PATH, autograd_gen_dir, 'tools/autograd')
  File "/home/kurtamohler/development/pytorch-bmm/tools/autograd/gen_autograd.py", line 234, in gen_autograd_python
    out, aten_decls + deprecated, template_path)
  File "/home/kurtamohler/development/pytorch-bmm/tools/autograd/gen_python_functions.py", line 173, in gen_py_torch_functions
    env = create_python_bindings(py_torch_functions, is_python_method=False, is_module=False)
  File "/home/kurtamohler/development/pytorch-bmm/tools/autograd/gen_python_functions.py", line 192, in create_python_bindings
    py_methods.append(method_impl(name, overload_decls, is_python_method, is_module))
  File "/home/kurtamohler/development/pytorch-bmm/tools/autograd/gen_python_functions.py", line 884, in method_impl
    grouped = group_overloads(declarations, is_python_method)
  File "/home/kurtamohler/development/pytorch-bmm/tools/autograd/gen_python_functions.py", line 1046, in group_overloads
    x, list(dictionary.keys())))
RuntimeError: 'base' not in dictionary for bmm(Tensor input, Tensor mat2, bool deterministic). keys are ['out', 'signature']
```

I've been trying different things to get it working, but can't figure out it.

ezyang replied:
OK, looks like the codegen doesn't actually understand how to deal with this :/ Optional positional arg it is, I suppose. I'll file an issue about this.

ezyang replied:
Filed at https://github.com/pytorch/pytorch/issues/33547

kurtamohler replied:
Cool thanks!

ezyang commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -708,4 +714,271 @@ Tensor _sparse_sum_backward_cuda(const Tensor& grad_, const SparseTensor& input_
+  // TODO: I'm not sure if this is necessary or not. Does at::resize_()
+  //    ensure contiguity?
+  if (!result.is_contiguous()) {
+    result.copy_(result.clone(at::MemoryFormat::Contiguous));
```
Yeah, I'm pretty sure this doesn't do what you want...

ezyang commented:
@pearu @nikitaved How about you guys do the first pass reviewing the algorithm? I can help with more framework review but I'd like you guys to do the bulk of the actual algorithm review.

kurtamohler committed:
Use temp result buffer if the given one is not contiguous

kurtamohler committed:
Make 'deterministic' flag a c10::optional

kurtamohler commented:
Below are charts of sparse-dense bmm's performance compared to a workaround function that someone posted to the original issue. My performance script is here: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/bmm-sparse-dense/bmm_perf.py

The "input cases" that the x-axis refers to are the input size combinations found in the same order in the table below the charts. "Pre-coalescing" means that before bmm() is called, I'm coalescing the sparse matrix outside the timed loop so that it does not have to be done inside the timed bmm() call. Measuring with and without the pre-coalescing step allows us to see how much of an impact coalescing has on the run time. Note that the way the workaround works, it does not need a coalesced sparse matrix, so it can skip that step. It looks like coalescing increases run time significantly in some cases but not in others.

The builtin methods are almost always faster than the workarounds, and CUDA is almost always faster than CPU.

![BMM sparse-dense performance (not pre-coalesced)](https://user-images.githubusercontent.com/8378787/74980770-b57ab700-53f6-11ea-82ac-6b528ecb7d90.png)

![BMM sparse-dense performance (pre-coalesced)](https://user-images.githubusercontent.com/8378787/74980785-b875a780-53f6-11ea-9972-ddddf84621ce.png)

**Input cases:**

num_matrices | squ_mat_elements | output_size | sparsity | nnz
-- | -- | -- | -- | --
10 | 100 | 10000 | 0.9 | 100
10 | 100 | 100000 | 0 | 1000
10 | 10000 | 1000000 | 0.999 | 100
10 | 10000 | 10000000 | 0.99 | 1000
10 | 10000 | 100000000 | 0.9 | 10000
10 | 10000 | 1000000000 | 0 | 100000
10 | 1000000 | 10000000000 | 0.999 | 10000
10 | 1000000 | 100000000000 | 0.99 | 100000
100 | 100 | 100000 | 0.9 | 1000
100 | 100 | 1000000 | 0 | 10000
100 | 10000 | 10000000 | 0.999 | 1000
100 | 10000 | 100000000 | 0.99 | 10000
100 | 10000 | 1000000000 | 0.9 | 100000
100 | 1000000 | 100000000000 | 0.999 | 100000
1000 | 100 | 1000000 | 0.9 | 10000
1000 | 100 | 10000000 | 0 | 100000
1000 | 10000 | 100000000 | 0.999 | 10000
1000 | 10000 | 1000000000 | 0.99 | 100000
10000 | 100 | 10000000 | 0.9 | 100000
10000 | 10000 | 1000000000 | 0.999 | 100000

kurtamohler commented:
I'm getting some CI errors related to JIT:
```
Feb 21 00:52:01 ======================================================================
Feb 21 00:52:01 ERROR [0.108s]: test_nested2 (__main__.EagerModePostTrainingQuantTest)
Feb 21 00:52:01 ----------------------------------------------------------------------
Feb 21 00:52:01 Traceback (most recent call last):
Feb 21 00:52:01   File "test_quantization.py", line 195, in test_nested2
Feb 21 00:52:01     checkQuantized(model)
Feb 21 00:52:01   File "test_quantization.py", line 193, in checkQuantized
Feb 21 00:52:01     self.checkScriptable(model, self.calib_data)
Feb 21 00:52:01   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_quantization.py", line 140, in checkScriptable
Feb 21 00:52:01     self._checkScriptable(orig_mod, traced, calib_data, check_save_load)
Feb 21 00:52:01   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_quantization.py", line 144, in _checkScriptable
Feb 21 00:52:01     self._checkModuleCorrectnessAgainstOrig(orig_mod, script_mod, calib_data)
Feb 21 00:52:01   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_quantization.py", line 161, in _checkModuleCorrectnessAgainstOrig
Feb 21 00:52:01     scripted_output = test_mod(inp)
Feb 21 00:52:01   File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 550, in __call__
Feb 21 00:52:01     result = self.forward(*input, **kwargs)
Feb 21 00:52:01   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 90, in prof_meth_call
Feb 21 00:52:01     return prof_callable(meth_call, *args, **kwargs)
Feb 21 00:52:01   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 84, in prof_callable
Feb 21 00:52:01     return callable(*args, **kwargs)
Feb 21 00:52:01 RuntimeError: 
Feb 21 00:52:01 Couldn't find an operator for aten::bmm(Tensor self, Tensor mat2) -> Tensor. Do you have to update a set of hardcoded JIT ops?
```
I'm not really sure what I need to do.

ezyang commented:
@kurtamohler Alright, it's this problem. Can you split bmm into two functions, `bmm(Tensor, Tensor)` and `_bmm(Tensor, Tensor, bool deterministic)`? Then have bmm just call `_bmm` with the deterministic flag set, but define autograd separately on `bmm` and `_bmm`. I think that will suffice to make this error go away.

cc'ing @eellison if you have a better protocol. (I am aware that we could also just bash this out by finding the schema string in JIT and updating it, but I kind of don't want to do that here; I feel the JIT doesn't want to see the determinism flag.)

eellison commented:
@ezyang @kurtamohler is the error is in shape analysis because `aten::bmm(Tensor self, Tensor mat2) -> Tensor` doesn't exist ? If that's the case, i would say just update it and hope that we are done deprecating it completely soon (cc @Krovatkin). Otherwise I'm not sure.

kurtamohler committed:
Attempt to fix 'bmm' JIT errors

kurtamohler commented:
Alright well I'm not sure if what I just changed fixed the issue or not. I haven't figured out how to manually run the failing test. So I'll just let CI run it.

kurtamohler commented:
Yesterday I discovered that my CPU implementation has a flaw in its method of searching the 3-D sparse tensor's indices for each 2-D matrix. Apparently, depending on how you create the sparse tensor, the tensor of indices can be laid out in either row-major or column-major order in memory. Looks like the `dense_tensor.to_sparse()` method gives the opposite index matrix ordering than what you get if you create a sparse matrix directly. My implementation was only handling one of these cases, and the following gave the wrong result:
```
a = torch.rand([2,2,2]).to_sparse()
b = torch.rand([2,2,2])
a.bmm(b)
```
Here is an example of how two different sparse tensors' index matrices can have different strides:
```
>>> b = torch.sparse.FloatTensor(torch.LongTensor([[0,0,1],[0,1,1],[0,1,1]]),torch.FloatTensor([1,2,3]),torch.Size([2,2,2])).coalesce()
>>> b.indices().stride()
(3, 1)
>>> b = b.to_dense().to_sparse()
>>> b.indices().stride()
(1, 3)
```

I have a working fix to my search function that takes this into account, and I will push it shortly.

kurtamohler committed:
Fix binary search to handle non-zero strides

Add test cases to verify this works correctly

kurtamohler committed:
Fix flake errors

ezyang commented:
re the hip errors (cc @iotamudelta )

```
22:43:07 /var/lib/jenkins/workspace/aten/src/ATen/native/sparse/hip/SparseHIPTensorMath.hip:793:1: error: unknown type name 'hipDataType_t'
22:43:07 hipDataType_t getTensorCudaDataType(Tensor self) {
22:43:07 ^
22:43:07 /var/lib/jenkins/workspace/aten/src/ATen/native/sparse/hip/SparseHIPTensorMath.hip:794:3: error: unknown type name 'hipDataType_t'
22:43:07   hipDataType_t cuda_data_type;
22:43:07   ^
22:43:07 /var/lib/jenkins/workspace/aten/src/ATen/native/sparse/hip/SparseHIPTensorMath.hip:797:24: error: use of undeclared identifier 'hipR32F'
22:43:07       cuda_data_type = hipR32F;
22:43:07                        ^
22:43:07 /var/lib/jenkins/workspace/aten/src/ATen/native/sparse/hip/SparseHIPTensorMath.hip:800:24: error: use of undeclared identifier 'hipR64F'
22:43:07       cuda_data_type = hipR64F;
22:43:07                        ^
22:43:07 /var/lib/jenkins/workspace/aten/src/ATen/native/sparse/hip/SparseHIPTensorMath.hip:880:3: error: unknown type name 'cusparseSpMMAlg_t'
22:43:07   cusparseSpMMAlg_t mm_alg = deterministic ? CUSPARSE_COOMM_ALG2 : CUSPARSE_COOMM_ALG1;
22:43:07   ^
22:43:07 /var/lib/jenkins/workspace/aten/src/ATen/native/sparse/hip/SparseHIPTensorMath.hip:880:46: error: use of undeclared identifier 'CUSPARSE_COOMM_ALG2'
22:43:07   cusparseSpMMAlg_t mm_alg = deterministic ? CUSPARSE_COOMM_ALG2 : CUSPARSE_COOMM_ALG1;
22:43:07                                              ^
22:43:07 /var/lib/jenkins/workspace/aten/src/ATen/native/sparse/hip/SparseHIPTensorMath.hip:880:68: error: use of undeclared identifier 'CUSPARSE_COOMM_ALG1'
22:43:07   cusparseSpMMAlg_t mm_alg = deterministic ? CUSPARSE_COOMM_ALG2 : CUSPARSE_COOMM_ALG1;
22:43:07                                                                    ^
22:43:07 /var/lib/jenkins/workspace/aten/src/ATen/native/sparse/hip/SparseHIPTensorMath.hip:899:11: error: unknown type name 'hipDataType_t'
22:43:07           hipDataType_t cuda_data_type = getTensorCudaDataType(mat2_contig);
22:43:07           ^
22:43:07 /var/lib/jenkins/workspace/aten/src/ATen/native/sparse/hip/SparseHIPTensorMath.hip:904:11: error: unknown type name 'cusparseSpMatDescr_t'; did you mean 
```

for now you should just ifdef out the code in the HIP case and say that it's not supported on HIP

iotamudelta commented:
It looks like either missing features in ROCm or (more likely) a mishipification. Agreed with @ezyang that this can be ifdef'd out for now, we'll have a look what's up there and may also ping you if we need some input?

kurtamohler commented:
Alright, thanks for letting me know. I'll put in the ifdef.

kurtamohler committed:
Add ifdef checks for HIP, plus message that it's not supported

nikitaved commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
+  }
+
+  return mid_ind;
+}
```
Since it is the `rightmost` variant, would it be more efficient to return `right_ind`? The logic seems like if `right_ind` already hits the rightmost, the method will not return until it is hit from the left...

kurtamohler replied:
I think you're right. The logic would be simpler that way. I'll change it and see how it affects performance.

kurtamohler replied:
Actually, most of the time when this search function is called, if the number of non-zero elements per matrix is roughly equal, the index we're searching for is going to be somewhere on the very left side of the array. The caller of this function is searching for the matrix indices in order, and it's shrinking the search area each time by starting from the right of the previously-found position. So its probably more likely that mid_ind will hit the position we're looking for before right_ind does. And anyway, this search is usually going to be a very small portion of the total runtime of bmm. I tried a basic linear search and the run time difference was very small, only about 1% slower in most of my test cases.

nikitaved commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -708,4 +714,290 @@ Tensor _sparse_sum_backward_cuda(const Tensor& grad_, const SparseTensor& input_
+
+  int64_t mat_el_begin_idx = 0;
+  size_t* workspace_buffer_sizes = new size_t[num_matrices];
+  void** workspace_buffers = new void*[num_matrices];
```
You sure there would not be any errors in-between to cause memory leaks?

kurtamohler replied:
Good point, if an exception is thrown, then the deletions at the end of the function wouldn't get run. I could put the array of buffers on the stack instead. But the buffers themselves have to be freed with cudaFree(), so I guess I need to use a try/catch to handle that case.

nikitaved replied:
Maybe I am wrong, but I am not sure you actually allocate this memory on gpu...

kurtamohler replied:
Yep `std::vector` is probably what I'll use.

ezyang replied:
Is the reason you need `num_matrices` workspace buffers because you run all of the individual unbatched sparse matrix multiplies in parallel?

kurtamohler replied:
Not exactly. From what I understand, each cuda function call gets sent to the gpu serially (correct me if I'm wrong). Avoiding the cudaFree between each matrix operation gave a measurable speedup. I'm not 100% sure why that is, and I don't recall how much speedup.

ezyang replied:
You're allocating and deallocating cuda manually; that's REALLY slow. It would be much better to get the scratch buffers by allocating a regular old tensor and then the memory will come from our caching allocator. But isn't the buffer the same size every call? So I don't think you need to allocate one buffer for everything.

kurtamohler replied:
The buffer might not be the same size each time, I'm not sure. It could potentially be different if the cusparseSpMM algorithm needs less space for fewer non-zero elements, for instance. But it's probably a better approach to reuse the same buffer and just increase its size only when needed.

ezyang replied:
Yes. In particular, this will reduce the maximum memory residency since you don't preallocate every buffer.

kurtamohler replied:
Done

nikitaved reviewed (approved):
LGTM. Thank you, @kurtamohler !

ezyang commented:
the SIGBUS looks related to this PR:

```

Mar 11 22:42:37 Running test_sparse ... [2020-03-11 22:42:37.452308]
Mar 11 22:42:39 
Mar 11 22:42:39 Running tests...
Mar 11 22:42:39 ----------------------------------------------------------------------
Mar 11 22:42:39 /opt/conda/lib/python3.6/site-packages/torch/tensor.py:746: UserWarning: The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad attribute won't be populated during autograd.backward(). If you indeed want the gradient for a non-leaf Tensor, use .retain_grad() on the non-leaf Tensor. If you access the non-leaf Tensor by mistake, make sure you access the leaf Tensor instead. See github.com/pytorch/pytorch/pull/30531 for more informations.
Mar 11 22:42:39   warnings.warn("The .grad attribute of a Tensor that is not a leaf Tensor is being accessed. Its .grad "
Mar 11 22:42:42 ..........Traceback (most recent call last):
Mar 11 22:42:42   File "test/run_test.py", line 674, in <module>
Mar 11 22:42:42     main()
Mar 11 22:42:42   File "test/run_test.py", line 667, in main
Mar 11 22:42:42     raise RuntimeError(message)
Mar 11 22:42:42 RuntimeError: test_sparse failed! Received signal: SIGBUS
```

ezyang commented:
For future reference, it'll be helpful if you describe the algorithm in the PR description. (I think it's laid out in the issue / comments, but the PR description will be the most obvious landing point for future people who want to find out more about the PR.)

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -646,13 +646,30 @@
+    SparseCUDA: bmm_sparse_cuda
+  supports_named_tensor: True
+
+- func: _bmm(Tensor self, Tensor mat2, bool deterministic) -> Tensor
```
In the issue we filed, we concluded that the kwargs didn't work because you didn't update the out variant too when you did it. So let's try kwarging this again.

kurtamohler replied:
Oh I think I misunderstood that. I'll try that.

kurtamohler replied:
What is the proper way to make the out variant use TensorOptions? I tried this:
```
- func: bmm.out(Tensor self, Tensor mat2, *, bool? deterministic=False, Tensor(a!) out) -> Tensor(a!)
```
But that apparently just turns deterministic into an `optional<bool>` for the out variant, even though it's a TensorOptions in the regular variant. When I try to build, I get this:
```
../aten/src/ATen/native/LinearAlgebra.cpp:313:61: error: could not convert 'options' from 'c10::TensorOptions' to 'c10::optional<bool>'
   return at::native::bmm_out_cpu(result, self, mat2, options);
```
I searched for other existing out functions that use TensorOptions, and I don't think there are any: `grep -R TensorOptions aten/src/ATen/native | grep '_out'`

ezyang replied:
Why do you want the out variant to use TensorOptions? Thi sisn't a factory function and I don't feel like that should happen here. Can you post the entire diff you did, and the auto-generated code in that case?

kurtamohler replied:
Apparently I still don't understand what I need to do. I thought that in order to make the kwargs work, I need to make these changes in native_functions.yaml:
```
- func: bmm(Tensor self, Tensor mat2, *, bool? deterministic=False) -> Tensor
...
- func: bmm.out(Tensor self, Tensor mat2, *, bool? deterministic=False, Tensor(a!) out) -> Tensor(a!)
```

After that, I thought I just needed to add a TensorOptions argument to the end of all the bmm and bmm_out function implementations.

What do I need to do instead?

kurtamohler replied:
Ok it seems like I needed to add optional<bool> to all of the bmm function implementations. That seems to work. I'm not sure why I thought I needed TensorOptions.

ezyang replied:
Great!

ezyang commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
+  SparseTensor self_coalesced;
+
+  if (!self.is_coalesced()){
+    self_coalesced = self.coalesce();
```
Huh, can't you just say `self_coalesced = self.coalesce()` unconditionally?

kurtamohler replied:
The `coalesce()` call takes a pretty long time, even if the tensor is already coalesced. I don't remember how much time, but it was pretty dramatic in some cases.

ezyang replied:
What do you mean by "long time"? 

```
SparseTensor coalesce_sparse_cpu(const SparseTensor& self) {
  AT_ASSERT(self.defined());
  TORCH_INTERNAL_ASSERT(at::impl::variable_excluded_from_dispatch());
  AT_ASSERT(self.is_sparse());
  
  if (self.is_coalesced()) {
    return self;
  }
```

and

```
SparseTensor coalesce_sparse_cuda(const SparseTensor& self) {
  int64_t nnz = self._nnz();
  if (self.is_coalesced()) {
    return self;
  }
```

So at most is should just be the cost of another dynamic dispatch.

ezyang replied:
So if it is actually taking a long time we should figure out why.

kurtamohler replied:
Ah you're right. I ran some direct measurements to compare whether my explicit `is_coalesced()` check saves any time. My claim that it's "pretty dramatic in some cases" was way off. There is a small speedup in most cases, but it's minute, usually around 1% or less. My guess is that this tiny speedup is coming from the fact that my explicit check avoids the three assertion checks in coalesce_sparse_cpu. I'll get rid of my explicit check since the gain is so small.


  |   |   |   |   | without checking |   | with checking |   | speedup with checking |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
device | num_matrices | squ_mat_elements | sparsity | workaround | not pre-coalesced | pre-coalesced | not pre-coalesced | pre-coalesced | not pre-coalesced | pre-coalesced
cpu | 10 | 100 | 0.9 | 0.001316 | 0.0006726 | 0.0004452 | 0.00067 | 0.0004418 | 1.003880597 | 1.00769579
cpu | 10 | 10000 | 0.999 | 0.001968 | 0.001083 | 0.0008478 | 0.0010672 | 0.0008358 | 1.014805097 | 1.014357502
cpu | 10 | 10000 | 0.99 | 0.002033 | 0.0014168 | 0.0009008 | 0.001391 | 0.0008846 | 1.018547807 | 1.018313362
cpu | 10 | 10000 | 0.9 | 0.002687 | 0.0048152 | 0.0013868 | 0.0048 | 0.0013716 | 1.003166667 | 1.011081948
cpu | 100 | 100 | 0.9 | 0.012417 | 0.004119 | 0.0036536 | 0.0041064 | 0.0036308 | 1.003068381 | 1.006279608
cpu | 100 | 10000 | 0.999 | 0.021109 | 0.0089946 | 0.0098808 | 0.0085026 | 0.0100254 | 1.057864653 | 0.9855766353
cpu | 100 | 10000 | 0.99 | 0.020283 | 0.0119564 | 0.0085836 | 0.0119212 | 0.0085624 | 1.002952723 | 1.002475941

ezyang replied:
Great. This might mean we should maybe optimize `coalesce_sparse_Cpu` a little, but that can always be something done later.

ezyang commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
+    // If there are sparse matrices at the beginning or end that
+    // have all zero elements, we need to zero out the result matrix.
+    if ((cur_mat_num < start_mat_num) || (mat_el_begin_idx >= nnz)) {
+      result[cur_mat_num].zero_();
```
I'm observing that you're doing a lot of tensor allocation inside the batch loop. This may be reducing the performance you're getting out of this. Something to think of if you're looking at perf on the cpu impl.

ezyang commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
+    // correspond to the current matrix number. We already know
+    // where the current matrix begins, so we just need to find
+    // the end. The search excludes everything to the left of
+    // the starting point, for best performance
```
Don't have to do it in this PR, but since you need to find *all* matix splits, I wonder if it wouldn't be better to do a single computation that precomputes all of the splits into a buffer, instead of repeatedly binary searching inside the batch loop.

kurtamohler replied:
If there's a performant way to do that, sure. Right now, the performance of finding the matrix splits is on the order of `O(num_matrices * ln(num_matrices * nnz_per_matrix))` (actually a bit better than that, since we're starting each subsequent search from the end of the previous search result). If we can get something better by somehow combining all the searches into one, then we should do that. But I don't really have a great idea how at the moment.

ezyang replied:
Well, the shitty algorithm is to just iterate over every entry one by one and record a split whenever it happens. That's `O(nnz)` but you have much better cache locality.

kurtamohler replied:
Linear iteration might not be so bad depending on the circumstance. It depends on how much faster (in terms of latency per access) that a linear access pattern is compared to a binary search access pattern, the caller's ratio of nonzeros versus number of matrices, and the total number of nonzeros.

At my previous job where I did competitive analysis of cache/memory performance on smartphone CPUs, it was common for a random memory or L3 cache access latency to be around 4-12 times slower than linear access latency (here's a memory latency graph of the Apple A13: https://www.anandtech.com/show/14892/the-apple-iphone-11-pro-and-max-review/3 compare the "linear chain" and "full random" lines on the chart), since prefetchers are good at handling linearity. I think we can say that a binary search's access pattern will behave fairly similarly to a random access (although maybe a little bit lower latency) as far as the prefetcher is concerned. When you get down to L2 the latency is about twice as fast for linear accesses versus random, and in L1, the access pattern makes no difference at all.

So a linear search will tend to be faster than the binary search if the total number of accesses multiplied by the access pattern latency is lower than that of the binary search. If we look at the best possible case where linear access could win, we'd look at the case where our total number of non-zeros is too large to fit within L2 and each linear access is 12 times faster than each binary search access. Applying this factor to the big-O runtimes of the two methods, we'd say that the linear search is preferable if `N < 12 * M * log2(N)`, where N is the total number of non-zeros and M is the number of matrices. If our L2 is 4 MB and we're using 32-bit floats, then N must be significantly more than 1,000,000 to get the 12x linear speedup--let's say we need `N > 1,000,000`. If we plug this into Wolfram Alpha: `N < 12 * M * log2(N), where 1000000 < N < 10000000 and 1 < M < 10000000`, we see that the inequality is true in almost the whole space (paying only attention to the part of the plot where N >= M https://www.wolframalpha.com/input/?i=N+%3C+12+*+M+*+log2%28N%29%2C+where+1000000+%3C+N+%3C+10000000+and+1+%3C+M+%3C+10000000), meaning that the linear access pattern actually could be faster in most cases (binary search is better if N/M is more than about 240). Even if the linear pattern does not give us any speedup, the binary search is only useful if the ratio of N/M is more than about 20: https://www.wolframalpha.com/input/?i=N+%3C+1+*+M+*+log2%28N%29%2C+where+1+%3C+N+%3C+100000+and+1+%3C+M+%3C+100000

But I don't know what a typical N/M ratio would be, so it's hard to know for sure if my reasoning here is realistic.

kurtamohler replied:
For what it's worth, I tried running a linear search in the CPU implementation. The performance was barely any different than the binary search:


num_matrices | squ_mat_elements | sparsity | nnz | builtin | builtin-coalesced
-- | -- | -- | -- | -- | --
10 | 100 | 0.9 | 100 | 0.9993763642 | 0.9941944848
10 | 100 | 0 | 1000 | 0.9943479171 | 0.9892473118
10 | 10000 | 0.999 | 100 | 1.000578369 | 0.9982660391
10 | 10000 | 0.99 | 1000 | 0.9908002336 | 0.965602518
10 | 10000 | 0.9 | 10000 | 1.01888459 | 0.9785580659
10 | 10000 | 0 | 100000 | 0.9809957734 | 0.929685379
10 | 1000000 | 0.999 | 10000 | 1.042729942 | 1.019447861
10 | 1000000 | 0.99 | 100000 | 1.025178905 | 1.00487879
100 | 100 | 0.9 | 1000 | 0.9959922814 | 0.9865570492
100 | 100 | 0 | 10000 | 0.9971276159 | 0.9865919064
100 | 10000 | 0.999 | 1000 | 0.9997310579 | 0.9952899695
100 | 10000 | 0.99 | 10000 | 0.9963695038 | 0.9998594881
100 | 10000 | 0.9 | 100000 | 0.9903190667 | 0.9664209664
100 | 1000000 | 0.999 | 100000 | 1.019451 | 1.047661575
1000 | 100 | 0.9 | 10000 | 0.9977529149 | 0.9945412865
1000 | 100 | 0 | 100000 | 0.9855607651 | 0.9915879947
1000 | 10000 | 0.999 | 10000 | 1.002447786 | 1.002210358
1000 | 10000 | 0.99 | 100000 | 1.049603886 | 1.04982573
10000 | 100 | 0.9 | 100000 | 0.9992745346 | 0.9982321546
10000 | 10000 | 0.999 | 100000 | 1.01204083 | 1.011304983

So maybe I should just use a linear search since the code is simpler. But again I don't know if my sizing parameters are realistic or not. What do you think?

ezyang commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
+      Tensor sparse_values = values.slice(0, mat_el_begin_idx, mat_el_end_idx);
+      int64_t sparse_nnz = mat_el_end_idx - mat_el_begin_idx;
+
+      AT_DISPATCH_ALL_TYPES(
```
Similarly, prefer not to repeatedly dispatch inside of the batch loop.

kurtamohler replied:
Done

ezyang commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
+Tensor& bmm_out_sparse_cpu(Tensor& result, const SparseTensor& self, const Tensor& mat2) {
+  TORCH_CHECK(!mat2.is_sparse(), "bmm_sparse: Tensor 'mat2' must be dense");
+
+  TORCH_CHECK(self.dense_dim() == 0, "bmm_sparse: Tensor 'self' must have 0 dense dims, but has ", self.dense_dim());
```
> Regarding how to define a batch of sparse matrices, we could accept a 3-D sparse tensor, just like how we're accepting a 3-D dense tensor. If we have a set of N 2-D tensors of identical shapes, the caller could torch.stack() them all into one sparse tensor.

I think this is probably right from an ease of API use perspective. When I think of blas-y APIs, they often let you pass in transpositions directly in the API, and that would have let you express a situation where you do pass in one dense dim at the end (which is the only thing sparse supports). I wonder if anyone will ask for this. (No action needed on this comment.)

ezyang commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
+
+  LongTensor indices_dim0 = indices[0];
+  int64_t indices_dim0_stride = indices_dim0.stride(0);
+  int64_t* indices_dim0_ptr = indices_dim0.data_ptr<int64_t>();
```
As the dimensionality and data type are known statically, you can use `TensorAccessor` for a much nicer, operator[] based API here.

kurtamohler replied:
Done

ezyang commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
+  int64_t dim_k = mat2.size(2);
+
+  Scalar beta = 0;
+  Tensor t = at::ones({self_coalesced.size(1)});
```
Oof. Really?

kurtamohler replied:
I might be able to avoid the requirement to provide this to `s_addmm_out_sparse_dense_cuda_worker`, let me see.

kurtamohler replied:
I just decided to make this tensor uninitialized, since this tensor won't even be looked at since beta = 0.

ezyang commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
+  ) {
+    // If there are sparse matrices at the beginning or end that
+    // have all zero elements, we need to zero out the result matrix.
+    if ((cur_mat_num < start_mat_num) || (mat_el_begin_idx >= nnz)) {
```
So, if I understand correctly, if there's a big stack of "zero" matrices at the end, we're just going to loop through them one by one and repeatedly call `zero_`. I wonder if it wouldn't be smarter to just up front zero the result tensor, and then walk down the index splits (which now itself can be sparse). Well, it's probably not worth it as I doubt people are going to do batched sparse matmul with a bunch of zero matrices, that's dumb lol.

kurtamohler replied:
Right, no one should really be using so many zero matrices, but this function should still support that. In most cases it will be more performant to avoid always zeroing out the entire thing up front.

ezyang commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
+  return bmm_out_sparse_cpu(result, self, mat2);
+}
+
+// Search a sorted strided array for the rightmost instance of a value.
```
This comment is a little confusing; what if the value doesn't exist?

ezyang replied:
A quick peek in the contents reveals that you return -1 if it is not found. What if `scalar_t` is an unsigned type? Answer: you only use this function at one specialization, `int64_t`.

kurtamohler replied:
I'm not actually checking for -1 anywhere though, I just figured I should assign it something consistent, rather than just returning garbage. The `found` bool reference is what the caller checks. I'll amend the comment to mention that.

ezyang commented on aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu:
```diff
@@ -708,4 +714,290 @@ Tensor _sparse_sum_backward_cuda(const Tensor& grad_, const SparseTensor& input_
+  }
+
+  // Dense matrices have to be contiguous for cusparseSpMM to work
+  const Tensor mat2_contig = mat2.is_contiguous() ? mat2 : mat2.clone(at::MemoryFormat::Contiguous);
```
`mat2.contiguous()` should be good enough her

kurtamohler committed:
Remove _bmm and _bmm_out, consilidate into bmm and bmm_out

kurtamohler committed:
Remove explicit is_coalesced() check

kurtamohler committed:
Simplify tensor contiguous call

kurtamohler committed:
Clean up commented code

kurtamohler committed:
Clarify search function description return by reference

kurtamohler committed:
Avoid initializing dummy tensor that is required for s_addmm_out_sparse_dense_worker

kurtamohler committed:
Move AT_DISPATCH outside of loop

kurtamohler committed:
Use TensorAccessor rather than a pointer to matrix indices

kurtamohler committed:
Use only one workspace buffer in bmm sparse-dense cuda

kurtamohler committed:
Move Scalar::to calls outside of loop

ezyang commented:
Given 

```
Mar 27 00:59:40 RuntimeError: 
Mar 27 00:59:40 Couldn't find an operator for aten::bmm(Tensor self, Tensor mat2) -> Tensor. Do you have to update a set of hardcoded JIT ops?
```

Either you have to update the schema strings, or I guess we have to do the same thing we did in the other diff where we split it into `bmm` (no kwarg) and `_bmm` (with the kwarg)

ezyang commented:
other pr being https://github.com/pytorch/pytorch/pull/34586

kurtamohler committed:
Change bmm schema in JIT shape analysis

kurtamohler committed:
Attempt to fix bmm deterministic kwarg

kurtamohler commented:
@ezyang, I tried to mimic the bool kwarg from #34586, but I'm getting JIT issues:
```
$ python test/test_jit.py TestJitGeneratedAutogradCPU.test_bmm_cpu
F
======================================================================
FAIL: test_bmm_cpu (__main__.TestJitGeneratedAutogradCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/work/kurtamohler/development/pytorch-bmm/torch/testing/_internal/common_device_type.py", line 207, in instantiated_test
    return test(self, device_arg)
  File "test/test_jit.py", line 18797, in do_test
    check(name)
  File "/work/kurtamohler/development/pytorch-bmm/torch/testing/_internal/jit_utils.py", line 614, in wrapper
    fn(*args, **kwargs)
  File "test/test_jit.py", line 18756, in check
    self.assertAutodiffNode(traced_fn.last_graph, should_autodiff_node, autodiff_nodes, fusible_nodes)
  File "/work/kurtamohler/development/pytorch-bmm/torch/testing/_internal/jit_utils.py", line 284, in assertAutodiffNode
    self.assertEqual(should_autodiff_node, found_all_nonfusible_nodes and found_all_fusible_nodes)
  File "/work/kurtamohler/development/pytorch-bmm/torch/testing/_internal/common_utils.py", line 907, in assertEqual
    super(TestCase, self).assertEqual(x, y, message)
AssertionError: True != False : 

----------------------------------------------------------------------
Ran 1 test in 0.733s

FAILED (failures=1)
```

The failure is at `self.assertAutodiffNode(traced_fn.last_graph, should_autodiff_node, autodiff_nodes, fusible_nodes)` within the `add_autograd_test` function in `test_jit.py`. I'm not yet sure why.

kurtamohler commented:
Following the jit test into assertAutodiffNode:

```
def assertAutodiffNode(self, graph, should_autodiff_node, nonfusible_nodes, fusible_nodes):
        diff_nodes = graph.findAllNodes('prim::DifferentiableGraph')
        diff_subgraphs = [node.g('Subgraph') for node in diff_nodes]

        # For any non-fusible node, it must show up in one of the DifferentiableGraph.
        found_all_nonfusible_nodes = (len(diff_subgraphs) == 0 and len(nonfusible_nodes) == 0)\
            or all([any(g.findNode(n) is not None for g in diff_subgraphs) for n in nonfusible_nodes])
        ...
```

`diff_nodes` is empty, but `nonfusible_nodes = ['aten::bmm']`. I don't know enough yet about how autograd works to know how to fix this, does anyone have an idea?

ezyang commented:
Is your attempt pushed? I don't see a `_bmm` anywhere

kurtamohler commented:
> Is your attempt pushed? I don't see a `_bmm` anywhere

Yes, it's my most recent two commits: https://github.com/pytorch/pytorch/pull/33430/commits/12a245bd084f973948a27732ee213065de439861
https://github.com/pytorch/pytorch/pull/33430/commits/a1eb13322afa5afaccfb17b4e39cd28c193cec56

I was trying to avoid using `_bmm`, since both your previous comments and #34586 seemed to indicate that it's possible. Should I just go back to using `_bmm` then?

ezyang commented:
You don't *have* to, but I misunderstood what your problem was because I thought that you were using the `_bmm` approach.

If you updated the schema directly, your bug is almost certainly because there is a schema string somewhere in the JIT that you haven't updated, OR if there is some JIT code that is manually creating a `bmm` call but hasn't filled in the optional argument. You'll have to find it, unfortunately. (Maybe it's easier to do `_bmm`. Sorry, I don't know the answer to this question. I just know that it would make the JIT problems go away. You'll have to add a `derivatives.yaml` for it...)

kurtamohler committed:
Revert back to using _bmm with deterministic argument

kurtamohler commented:
I couldn't figure out what was causing JIT issues, so I just went back to using `_bmm`.

kurtamohler commented:
I'm not sure what I need to do to fix the Windows cusparse errors.

kurtamohler commented:
Well looking at the build commands, the Windows job is using CUDA 10.1. That version of cuda does have the `cusparseCreateDnMat` function: https://docs.nvidia.com/cuda/archive/10.1/pdf/CUSPARSE_Library.pdf

I think `-lcusparse` isn't included in the build command, but I'm not completely sure. I don't really know what these `C:/Users/circleci/project/build/win_tmp/bin/randomtemp.exe` commands are, but here's the failing one:
```
C:/Users/circleci/project/build/win_tmp/bin/randomtemp.exe C:/Users/circleci/project/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu -c -o C:/Users/circleci/project/build/caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/sparse/cuda/./torch_cuda_generated_SparseCUDATensorMath.cu.obj -ccbin cl.exe -m64 -Dtorch_cuda_EXPORTS -DTORCH_CUDA_BUILD_MAIN_LIB -DTH_BLAS_MKL -D_OPENMP_NOFORCE_MANIFEST -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -D_CRT_SECURE_NO_DEPRECATE=1 -DWIN32_LEAN_AND_MEAN -DMAGMA_V2 -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DAT_PARALLEL_OPENMP=1 -Xcompiler ,\"/DWIN32\",\"/D_WINDOWS\",\"/GR\",\"/w\",\"/bigobj\",\"-openmp:experimental\",\"-DNDEBUG\",\"-DUSE_FBGEMM\",\"-DHAVE_AVX_CPU_DEFINITION\",\"-DHAVE_AVX2_CPU_DEFINITION\",\"/MD\",\"/O2\",\"/Ob2\",\"/DNDEBUG\",\"/w\",\"/bigobj\",\"-DNDEBUG\" -Xcompiler /w -w -DONNX_NAMESPACE=onnx_torch --use-local-env -gencode arch=compute_75,code=sm_75 -Xcudafe --diag_suppress=cc_clobber_ignored -Xcudafe --diag_suppress=integer_sign_change -Xcudafe --diag_suppress=useless_using_declaration -Xcudafe --diag_suppress=set_but_not_used -Xcompiler -MD --expt-relaxed-constexpr --expt-extended-lambda -Xcompiler /wd4819 -Xcompiler /wd4503 -Xcompiler /wd4190 -Xcompiler /wd4244 -Xcompiler /wd4251 -Xcompiler /wd4275 -Xcompiler /wd4522 -Wno-deprecated-gpu-targets --expt-extended-lambda -gencode arch=compute_75,code=sm_75 -DCUDA_HAS_FP16=1 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ -DNVCC "-IC:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/include" -IC:/Users/circleci/project/build/aten/src -IC:/Users/circleci/project/aten/src -IC:/Users/circleci/project/build -IC:/Users/circleci/project -IC:/Users/circleci/project/cmake/../third_party/googletest/googlemock/include -IC:/Users/circleci/project/cmake/../third_party/googletest/googletest/include -IC:/Users/circleci/project/third_party/protobuf/src -IC:/Users/circleci/project/build/win_tmp/mkl/include -IC:/Users/circleci/project/cmake/../third_party/benchmark/include -IC:/Users/circleci/project/third_party -IC:/Users/circleci/project/cmake/../third_party/eigen -IC:/Jenkins/Miniconda3/include -IC:/Jenkins/Miniconda3/lib/site-packages/numpy/core/include -IC:/Users/circleci/project/cmake/../third_party/pybind11/include -I/opt/rocm/hip/include -I/include -IC:/Users/circleci/project/cmake/../third_party/cub -IC:/Users/circleci/project/build/caffe2/contrib/aten -IC:/Users/circleci/project/third_party/onnx -IC:/Users/circleci/project/build/third_party/onnx -IC:/Users/circleci/project/third_party/foxi -IC:/Users/circleci/project/build/third_party/foxi -IC:/Users/circleci/project/build/win_tmp/magma/include -IC:/Users/circleci/project/third_party/ideep/mkl-dnn/include -IC:/Users/circleci/project/third_party/ideep/include "-IC:/Program Files/NVIDIA Corporation/NvToolsExt/include" -IC:/Users/circleci/project/build/caffe2/aten/src/TH -IC:/Users/circleci/project/aten/src/TH -IC:/Users/circleci/project/build/caffe2/aten/src/THC -IC:/Users/circleci/project/aten/src/THC -IC:/Users/circleci/project/aten/src/THCUNN -IC:/Users/circleci/project/aten/src/ATen/cuda -IC:/Users/circleci/project/build/caffe2/aten/src -IC:/Users/circleci/project/aten/../third_party/catch/single_include -IC:/Users/circleci/project/aten/src/ATen/.. -IC:/Users/circleci/project/build/caffe2/aten/src/ATen -IC:/Users/circleci/project/c10/cuda/../.. -IC:/Users/circleci/project/c10/../ -IC:/Users/circleci/project/caffe2/../torch/csrc/api -IC:/Users/circleci/project/caffe2/../torch/csrc/api/include -IC:/Users/circleci/project/build/third_party/ideep/mkl-dnn/include -IC:/Users/circleci/project/third_party/ideep/mkl-dnn/src/../include
```

ezyang commented:
cc @peterjc123. I agree this is pretty strange.

ezyang commented:
The functions you're using are not supported on Windows. From the manual:

```
Starting with CUDA 10.1, the cuSPARSE Library provides a new set of API to compute
sparse vector-vector (SpVV), matrix-vector (SpMV), and matrix-matrix multiplication
(SpMM), in addition to the existing legacy API. They are considered a preview feature i.e.
the data structures and APIs for them are subject to change and may not be compatible
with future releases.
LIMITATION: The generic APIs are currently available for all platforms except
Windows. Using these APIs in any other systems will result in compile-time or run-time
failures. Their support will be extended in the next releases.
```

https://docs.nvidia.com/cuda/archive/10.1/pdf/CUSPARSE_Library.pdf

ezyang commented:
Thus, you should ifdef out the functionality on windows and skip the tests on windows.

kurtamohler commented:
That paragraph from the manual says that Windows support will be expanded in future, so we should make sure we revisit this in the future when newer CUDA versions are released.

kurtamohler committed:
Remove bmm sparse-dense CUDA Windows support

ezyang commented:
BTW, when you have so many commits, consider not rebasing and force pushing in the future ;)

kurtamohler commented:
@ezyang , oh ok, what's the better way to do it?

kurtamohler commented:
I have a suspicion that the SIGBUS error showing up in `pytorch_linux_xenial_cuda10_2_cudnn7_py3_gcc7_test` might be due to the fact that I'm using `cudaMallocManaged`. How do I find out which GPU architecture that job is running on? Apparently anything earlier than Kepler doesn't have the unified memory feature that I was trying to use: https://stackoverflow.com/questions/20074101/would-it-be-possible-to-access-to-gpu-ram-from-cpu-cores-by-simple-pointer-in-a

ezyang commented:
You can SSH into the CircleCI job and debug it directly there (Rerun with ssh), does that help?

kurtamohler commented:
@ezyang , oh I didn't know that. Where do I find the address for ssh?

kurtamohler commented:
@ezyang, when I ssh into the CircleCI job, I can't even find any filenames related to pytorch on the machine. I get nothing when I run `find / -name '*torch*' 2>/dev/null`

What am I supposed to do?

ezyang commented:
You need to enter the docker image. `docker ps` to find the id, then `docker exec -it IMAGE_ID /bin/bash`

kurtamohler committed:
Use explicit cuda host-memory data transfer in bmm sparse-dense

kurtamohler committed:
Fix test_bmm_deterministic to allow for small difference between results

kurtamohler committed:
Avoid using TORCH_CUDASPARSE_CHECK for cuda malloc and memcpy

kurtamohler commented:
After looking in gdb, I found that the problem with the SIGBUS error in the `pytorch_linux_xenial_cuda10_2_cudnn7_py3_gcc7_test` job actually was caused by accessing a `cudaMallocManaged` array from the cpu, like I had suspected before. I'm not sure why it doesn't work, since the gpu for that job is supposed to have unified memory. Nevertheless, I switched to regular `cudaMalloc` with explicit device-to-host `cudaMemcpy`, and now that job is passing.

The Windows failure is real though, so I'll look into that.

kurtamohler committed:
Add printout of relative difference for CI debug

kurtamohler committed:
Disable test_bmm_deterministic for windows

kurtamohler committed:
Skip bmm deterministic error on Windows, and add Windows error test

kurtamohler committed:
Fix flake error

kurtamohler committed:
Fix indentation flake error

kurtamohler committed:
Fix Windows ifdef in bmm sparse-dense

kurtamohler committed:
Use TORCH_CHECK instead of TORCH_INTERNAL_ASSERT

kurtamohler commented:
Alright, I fixed my `#ifdef`s to include `_WIN64`. I was only checking `_WIN32` before. I also disabled `test_bmm_deterministic` for Windows and added a test to make sure that bmm sparse-dense cuda throws an error when it's run on Windows. The Windows CI job now passes.

peterjc123 commented:
> Alright, I fixed my `#ifdef`s to include `_WIN64`. I was only checking `_WIN32` before.

It's weird. `_WIN32` should also be defined in x64 platforms.

kurtamohler commented:
@peterjc123, I think you're right, `WIN32_` should be enough. I think I was using `TORCH_INTERNAL_ASSERT` incorrectly. I incorrectly thought that it was an unconditional version of `TORCH_CHECK` and only needed one argument (the error message). So, changing back to using `TORCH_CHECK` with the conditional set to false must have been the real reason why the error is now being thrown given the correct conditions. I think this might be a demonstration of why preprocessor macros should be avoided.

ezyang commented:
Drat, it looks like our internal version of cusparse is too old. Is there a way you can add macro ifdefs that appropriate test the version of cusparse before making API calls?

```
stderr: caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(807): warning: statement is unreachable
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(898): error: identifier "cusparseSpMMAlg_t" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(898): error: identifier "CUSPARSE_COOMM_ALG2" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(898): error: identifier "CUSPARSE_COOMM_ALG1" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "cusparseSpMatDescr_t" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "CUSPARSE_INDEX_32I" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "cusparseCreateCoo" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "cusparseDnMatDescr_t" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "CUSPARSE_ORDER_COL" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "cusparseCreateDnMat" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "cusparseDnMatDescr_t" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "CUSPARSE_ORDER_COL" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "cusparseCreateDnMat" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "cusparseSpMM_bufferSize" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "cusparseSpMM" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "cusparseDestroySpMat" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "cusparseDestroyDnMat" is undefined
caffe2/aten/src/ATen/native/sparse/cuda/SparseCUDATensorMath.cu(902): error: identifier "cusparseDestroyDnMat" is undefined
```

kurtamohler committed:
Throw error if bmm sparse-dense is run with CUDA less than 10.1

kurtamohler committed:
Fix bmm cuda version test skipIf conditions

kurtamohler commented:
Alright, I changed the ifdefs so that an error is thrown if the cuda version is less than 10.1. I also added a test to make sure the error is being thrown correctly and skip the other tests if less than cuda 10.1.

kurtamohler committed:
Fix cuda version comparison

kurtamohler commented:
Looks like the macos environment doesn't have the `torch._C._cuda_getCompiledVersion()` function that I was using to decide whether to skip tests. I changed it to use `torch.version.cuda` instead, hopefully that's available in all the testing environments.

kurtamohler commented:
Darnit, the version string comparison module I was using doesn't exist on macos and windows:
```
    from packaging import version
ModuleNotFoundError: No module named 'packaging'
```

kurtamohler committed:
Fix version string comparison

kurtamohler commented:
I removed the packaging module import and chose to use this method of comparison instead:
```
[int(x) for x in torch.version.cuda.split(".")] >= [10, 1]
```
Hopefully that's robust enough not to break.

kurtamohler committed:
Avoid version compare if torch.version.cuda is None

kurtamohler commented:
Wasn't robust enough. I think this will be.

kurtamohler committed:
Fix flake error

ezyang commented:
This looks like it was sufficient. Unfortunately it looks like we need to merge with master.

ezyang committed:
Merge remote-tracking branch 'origin/master' into bmm-sparse-dense-5672

kurtamohler commented:
@ezyang, do you know what caused Facebook Internal to fail?

ezyang commented:
it's fake, I think
